File: sonar-server/src/main/java/org/sonar/server/exceptions/Messages.java
/*
 * SonarQube, open source software quality management tool.
 * Copyright (C) 2008-2014 SonarSource
 * mailto:contact AT sonarsource DOT com
 *
 * SonarQube is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * SonarQube is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.server.exceptions;

import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.ListMultimap;
import org.sonar.api.utils.text.JsonWriter;

import java.util.Collection;
import java.util.List;

public class Messages {

  private ListMultimap<Message.Level, Message> messagesByLevel = ArrayListMultimap.create();

  public Messages add(Message message) {
    messagesByLevel.put(message.getLevel(), message);
    return this;
  }

  public Collection<Message> all() {
    return messagesByLevel.values();
  }

  public List<Message> getByLevel(Message.Level level) {
    return messagesByLevel.get(level);
  }

  public boolean hasLevel(Message.Level level) {
    return messagesByLevel.containsKey(level);
  }

  public void writeJson(JsonWriter json) {
    writeJson(json, "errors", Message.Level.ERROR);
    writeJson(json, "warnings", Message.Level.WARNING);
    writeJson(json, "infos", Message.Level.INFO);
  }

  private void writeJson(JsonWriter json, String label, Message.Level level) {
    List<Message> messages = messagesByLevel.get(level);
    if (!messages.isEmpty()) {
      json.name(label).beginArray();
      for (Message message : messages) {
        json.beginObject().prop("msg", message.getLabel()).endObject();
      }
      json.endArray();
    }
  }
}

File: server/sonar-web/src/main/js/api/messages.ts
/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { throwGlobalError } from '~sonar-aligned/helpers/error';
import { getJSON } from '~sonar-aligned/helpers/request';
import { post } from '../helpers/request';

export enum MessageTypes {
  GlobalNcd90 = 'GLOBAL_NCD_90',
  GlobalNcdPage90 = 'GLOBAL_NCD_PAGE_90',
  ProjectNcd90 = 'PROJECT_NCD_90',
  ProjectNcdPage90 = 'PROJECT_NCD_PAGE_90',
  BranchNcd90 = 'BRANCH_NCD_90',
  UnresolvedFindingsInAIGeneratedCode = 'UNRESOLVED_FINDINGS_IN_AI_GENERATED_CODE',
}

export interface MessageDismissParams {
  messageType: MessageTypes;
  projectKey?: string;
}

export function checkMessageDismissed(data: MessageDismissParams): Promise<{
  dismissed: boolean;
}> {
  return getJSON('/api/dismiss_message/check', data).catch(throwGlobalError);
}

export function setMessageDismissed(data: MessageDismissParams): Promise<void> {
  return post('/api/dismiss_message/dismiss', data).catch(throwGlobalError);
}
